common/kafka/kafka_consumer.cpp

```diff
@@ -367,46 +332,82 @@ bool KafkaConsumer::SeekInternal(
     LOG(INFO) << "Seeking partition: " << topic_partition->partition()
               << ", offset: " << topic_partition->offset();
 
+    std::pair<std::string, std::int32_t> topic_partition_pair =
+      std::make_pair(topic_partition->topic(), topic_partition->partition());
+
+    ConsumedOffset consumedOffset;
+    consumedOffset.offset = topic_partition->offset();
+    consumedOffset.timestamp = timestamp_ms;
+
     common::Stats::get()->Incr(
-        getFullStatsName(kKafkaConsumerSeek,
-                           {kafka_consumer_type_metric_tag_,
-                            "topic=" + topic_partition->topic(),
-                            "partition=" +
-                            std::to_string(topic_partition->partition())}));
-    if (!KafkaSeekWithRetry(consumer_.get(), *topic_partition)) {
+      getFullStatsName(kKafkaConsumerSeek,
+                       {kafka_consumer_type_metric_tag_,
+                        "topic=" + topic_partition->topic(),
+                        "partition=" +
+                        std::to_string(topic_partition->partition())}));
+    if (!KafkaSeekWithRetry(consumer_->getInstance().get(), *topic_partition)) {
       return false;
     }
+    consumedOffsets_[topic_partition_pair] = consumedOffset;
   }
   return true;
 }
 
-RdKafka::Message* KafkaConsumer::Consume(int32_t timeout_ms) const {
-  if (consumer_ == nullptr) {
-    return nullptr;
-  }
+RdKafka::Message* KafkaConsumer::Consume(int32_t timeout_ms) {
+  do {
+    if (consumer_ == nullptr || consumer_->getInstance() == nullptr) {
+      return nullptr;
+    }
 
-  auto* message = consumer_->consume(timeout_ms);
+    while (std::atomic_exchange(&reset_, false)) {
+      consumer_->resetInstance(&consumedOffsets_);
+    }
```

Comment:
This function is in the critical path for consuming kafka messages. Is it necessary to do a read-modify-write every time? Can we do just a read, instead? Or rely on the logic in lines 368->372 to reset the consumer?
I'm concerned about the performance impact of this (especially for non-PII cases where reset will never be set to True)
How do you feel about having another variable to indicate if this is a PII use case or not (or of the file watcher has been setup)? This variable will not change once initialized, and we could check that first before checking _reset.

Reply:
I modified this logic.. with a condition which check's if atomic operations are required at all.